### ISSUE TITLE
b stack buffer overrun detected when trying to setup an unusual configuration (windows)

### ISSUE THREAD

**Issue by Klaim** [build2/build2#38]
I didn't manage to reproduce this with a simpler case for now (lack of time, I'll try to do more soon). 
I don't think it's relative to the actual library used, but that's how I got there.

How to reproduce:
============

0. On Windows, with build2 v0.11.0, using msvc 16.1 (used for all the following and the build2 install).
1. build and install boost 1.69.0 somewhere on the disk (using `b2`) (or mayb just download a binary package? I didn't try that yet)
2. Add a pkg-config file for boost in the `lib/pkgconfig` directory of your boost install. Here is what I used: https://pastebin.com/7A01nCw7
3. `bdep new boostuser && cd boostuser`
4. Add ` depends: libboost >= 1.69.0` in `manifest`.
5. Create the configuration this way, just replace the paths to boost install (yep, this is weird): 
```
bdep init ?sys:libboost/* -C ../build-boostuser cc config.cxx.loptions="/LIBPATH:G:\tools\boost\boost_install\lib" config.cxx.poptions="/I G:\tools\boost\boost_install\include\boost-1_69"
```

This leads to an error because for some reason `?sys:libboost/*` is ignored even if put at the end of the line. (see #39 )

6. Insist again: `bdep init ?sys:libboost/*` => lead to the error

Observed:
=======

```
PS G:\tmp\test_build2_unpackaged\boostuser> bdep init ?sys:libboost/* -C ../build-boostuser cc config.cxx.loptions="/LIBPATH:G:\tools\boost\boost_install\lib" config.cxx.poptions="/I G:\tools\boost\boost_install\include\boost-1_69"   initializing in project G:\tmp\test_build2_unpackaged\boostuser\
created configuration G:\tmp\test_build2_unpackaged\build-boostuser\ 1 default,forwarded,auto-synchronized
error: unknown dependency libboost >= 1.69.0 of package boostuser
  info: repository dir+file:///G:/tmp/test_build2_unpackaged/boostuser appears to be broken
  info: or the repository state could be stale
  info: run 'bpkg rep-fetch' to update
info: while satisfying boostuser/0.1.0-a.0.19700101000000
PS G:\tmp\test_build2_unpackaged\boostuser> bdep init ?sys:libboost/*                                                initializing in project G:\tmp\test_build2_unpackaged\boostuser\
synchronizing:
  configure sys:libboost/*
  new boostuser/0.1.0-a.0.19700101000000
error: process b terminated abnormally: stack buffer overrun
```

With `--verbose 4` : https://pastebin.com/zGWukPx0

**Comment by karen-arutyunov** [comment]
Joël, the "How to reproduce" instructions require 'add depends: libboost >= 1.65.0 in the manifest' item, right?

**Comment by Klaim** [comment]
Yes, sorry I forgot to clarify: I added `depends: libboost >= 1.69.0`
I'll update the description.

**Comment by karen-arutyunov** [comment]
I believe that pointing to the real boost installation is not really required, though unable to reproduce the issue myself yet. Could you try to reproduce it changing paths in p.5 to refer to non-existent paths, for example:

```
bdep init ?sys:libboost/* -C ../build-boostuser cc config.cxx.loptions="/LIBPATH:K:\tools\boost\boost_install\lib" config.cxx.poptions="/I K:\tools\boost\boost_install\include\boost-1_69"
```
Does it crashes with "stack buffer overrun" every time you run this sequence of commands?

**Comment by Klaim** [comment]
It does not crash in that case:

```
PS G:\tmp\test_build2_unpackaged\boostuser> bdep init ?sys:libboost/* -C ../build-boostuser cc config.cxx.loptions="/LIBPATH:K:\tools\boost\boost_install\lib" config.cxx.poptions="/I K:\tools\boost\boost_install\include\boost-1_69"   initializing in project G:\tmp\test_build2_unpackaged\boostuser\
created configuration G:\tmp\test_build2_unpackaged\build-boostuser\ 1 default,forwarded,auto-synchronized
error: unknown dependency libboost >= 1.69.0 of package boostuser
  info: repository dir+file:///G:/tmp/test_build2_unpackaged/boostuser appears to be broken
  info: or the repository state could be stale
  info: run 'bpkg rep-fetch' to update
info: while satisfying boostuser/0.1.0-a.0.19700101000000
PS G:\tmp\test_build2_unpackaged\boostuser> bdep init ?sys:libboost/*
initializing in project G:\tmp\test_build2_unpackaged\boostuser\
synchronizing:
  configure sys:libboost/*
  new boostuser/0.1.0-a.0.19700101000000
error: unable to import target libboost%lib{boost_filesystem}
  info: use config.import.libboost command line variable to specify its project out_root
```

If I try with a `libextra` package:
 -  `bdep new libextra -t lib`  then I install it somewhere 
 - replace the boost line in the manifest by `depends: libextra >= 0.1.0`
Then I don't observe a crash:

```
PS G:\tmp\test_build2_unpackaged\boostuser> bdep init ?sys:libextra/* -C ../build-boostuser cc config.cxx.loptions="/LIBPATH:G:\tmp\test_build2_unpackaged\install\lib" config.cxx.poptions="/I G:\tmp\test_build2_unpackaged\install\include"                                                                                                                 initializing in project G:\tmp\test_build2_unpackaged\boostuser\
created configuration G:\tmp\test_build2_unpackaged\build-boostuser\ 1 default,forwarded,auto-synchronized
error: unknown dependency libextra >= 0.1.0 of package boostuser
  info: repository dir+file:///G:/tmp/test_build2_unpackaged/boostuser appears to be broken
  info: or the repository state could be stale
  info: run 'bpkg rep-fetch' to update
info: while satisfying boostuser/0.1.0-a.0.19700101000000
PS G:\tmp\test_build2_unpackaged\boostuser> bdep init ?sys:libextra/*
initializing in project G:\tmp\test_build2_unpackaged\boostuser\
synchronizing:
  configure sys:libextra/*
  new boostuser/0.1.0-a.0.19700101000000

```

I suspect that I'm doing something in the pkg-config file that I manually wrote for my boost install that triggers the error.

**Comment by Klaim** [comment]
> Does it crashes with "stack buffer overrun" every time you run this sequence of commands?

To be completely clear, the only case I found so far resulting this crash was with the repro case.
I didn't try yet to reproduce it on linux with the same pkg-config file though. I'll try that when I can.

**Comment by karen-arutyunov** [comment]
Thanks Joël,

If you could reproduce it on Linux it would be great. 

I also now realized from the diagnostics you provided lately that your buildfile is not a basic one created by `bdep new` but something more advanced, importing boost library at least. Could you also share this buildfile?

**Comment by Klaim** [comment]
Sorry about that, I indeed forgot that I added an import of boost filesystem and `#include <boost/filesystem.hpp>`.
I also found that so far I can't reproduce the issue on another windows computer (at home, but with a boost 1.70 built).

So let met get back next week on the repro case, I'll upload it on github and continue trying to find what triggers it exactly.

**Comment by Klaim** [comment]
I didn't manage to make a smaller repro case yet (with something else than boost and my pc file) but:

 - Here is the exact code I use that trigger the issue: https://github.com/Klaim/repro-case-buffer-overrun-with-boost/tree/master/boostuser
 - I found that removing/commenting the import `import libs += libboost%lib{boost_filesystem}` will avoid the stack overrun;
 - I found that naming my boost pc file from `libboost.pc` to `boost.pc` will avoid the stack overrun (the lib is just not found in this case).
 - I tried the changing the name of the libextra, to reproduce the problem but it doesn't stack overrun.

I'm trying to upload my boost install directory somewhere as I can't find another way to reproduce it so far.

**Comment by Klaim** [comment]
Another info: the boost install I have is actually an older version of msvc abi than the msvc version I'm using to test, and the pc file is probably incorrect as it does not specify which msvc, either 32 or 64 bit etc.
I'm building another version of boost with the same version of msvc than my test project, see if it changes something.

**Comment by Klaim** [comment]
The [repro case repository (master)](https://github.com/Klaim/repro-case-buffer-overrun-with-boost) is now setup with a fake pc file that triggers the issue at least on my dayjob windows machine. (no ndeed for any actual library, the pc file alone is enough)

On windows, just run `repro.bat`, it will first remove previously generated files then just do the `bdep init` sequence that trigger the error, using the fake pc file.

I managed to reproduce the issue on 2 different Windows 10 computers. To be sure I tried from different directories and that doesn't change the outcome.

Here are my other findings:

 - I still didn't manage to reproduce the issue on linux so far: I get no error at all (not even that it's still not finding the library), which I guess should be expected as the wrong info in the pc file are compile-time info;
 - I sometime didn't get the buffer overrun after modifying the .pc file, but I couldn't find what exactly trigger the issue in the file. For example at some point I realized that changing `G:\that\path\` to `G:\\that\\path` in the pc file will remove the error. However while making the repro case, that became false (I still get buffer overruns), but I'm not sure what I changed that triggered the error.

To me all these points seems to confirm that something happen when reading/interpreting/executing? the pc file and then b reach the buffer overrun state.

Maybe it's a compiler generated code (of `b`) that is wrong, so to clarify I get the issue with this version of `cl` (which I also used to build `build2`):

```
PS G:\tmp\repro-case-buffer-overrun-with-boost> cl                                                                   Microsoft (R) C/C++ Optimizing Compiler Version 19.21.27702.2 for x64
```

Tell me if you want me to try building build2 with an older version of cl.

Oh also, on one of the compilers I had a stage version of build2 which didn't have the buffer overrun issue (it got another error but I believe it was fixed), so this is a recent issue, maybe emerging from a recent fix.
